### ISSUE TITLE
Non panicking assertions

### ISSUE THREAD

**Issue by CodesInChaos** [Rust-for-Linux/linux#201]
By default Rust takes a fail fast approach, where failing assertions panic. Due to the big impact of a kernel panic, this approach [isn't quite as popular](https://lkml.org/lkml/2016/10/4/1) in the Linux kernel.

So it'd probably be a good idea to add helper functions which merely trigger a warning and return with an error. Unfortunately I couldn't find an error code for "internal error".

I think the most common functions/macros which could use a less lethal equivalent would be `panic!`, `assert!`, `assert_eq!`, `unwrap`, `expect`.

**Comment by Stargateur** [comment]
rust assert are not mean to "report error" if you assert something it's considered at a bug if this assert is not true, I fail to follow what you expect from a code that expect something to be true but isn't, the following code will probably fail or introduce security problem.

You can't trivially replace an unwrap or return an error in place of an assert. And sometime it's make sense to use them.

> Killing machines because somebody made an assumption that was wrong is not ok.
> Killing the machine is ok if we have a situation where there literally
is no other choice.

Well guess the solution is to disallow any code that include a panic. But this is not trivial.

Also, https://lkml.org/lkml/2016/10/4/337, look like the problem is that the assert was premature.

**Comment by nbdd0121** [comment]
Rust uses assertions in many places to ensure safety, e.g. to verify preconditions hold for unsafe code. These assertions must panic because killing kernel is better than a memory safety violation. In the cases you point to, it should be a WARN_ON rather than BUG_ON. And we should have some extra macros for that semantics, but they should not replace `assert` and `panic`.

Also, for `panic`, the return value is `!`, so you cannot return from it, otherwise it is UB. For `unwrap()` and `expect()`, the return value is `T`, so returning when the unwrapped object is `None` or `Err` simply makes no sense and would easily be UB.

**Comment by CodesInChaos** [comment]
I'm not talking about changing the existing methods, but adding similar ones, which trigger a warning and return a `KernelResult` you can propagate with `?`.

e.g. you could add a macro `assert_warn!`, which checks the condition, and returns early with an error from the surrounding function in case of an error.

**Comment by elpiel** [comment]
> Rust uses assertions in many places to ensure safety, e.g. to verify preconditions hold for unsafe code
@nbdd0121 I think this is still useful, especially if you can opt-in for this behavior.

This is somewhat achievable with [std::panic::set_hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) & [std:panic::catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)

There are caveats as well, though:
https://doc.rust-lang.org/std/panic/fn.catch_unwind.html#notes

> Note that this function might not catch all panics in Rust. A panic in Rust is not always implemented via unwinding, but can be implemented by aborting the process as well. This function only catches unwinding panics, not those that abort the process.
> 
> Also note that unwinding into Rust code with a foreign exception (e.g. an exception thrown from C++ code) is undefined behavior.

**Comment by nbdd0121** [comment]
> > Rust uses assertions in many places to ensure safety, e.g. to verify preconditions hold for unsafe code
>
> @nbdd0121 I think this is still useful, especially if you can opt-in for this behavior.
> 
> This is somewhat achievable with [std::panic::set_hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) & [std:panic::catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)

No, that's still panicking, and the control flow does not continue at assert, but at catch_unwind. Panic still happens, it's just not aborting.

Plus, we are 1) not using unwinding in the kernel, and 2) not using std in the kernel.

**Comment by ojeda** [comment]
> This is somewhat achievable with [std::panic::set_hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) & [std:panic::catch_unwind](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html)

Note that we are in `no_std` + `panic=abort` land. Even if we were in a normal application, playing around with unwinding is usually not a good idea.

**Comment by elpiel** [comment]
Thank you @ojeda & @nbdd0121 for the information. I will closely follow the issue as I'm interested in this.

**Comment by ojeda** [comment]
Closing this -- this is a very old issue, and nowadays in upstream we have different mechanisms we use (panicking, "warnings", debug asserts / erroneous behavior, etc.) depending on the situation, some named similarly to the C side.
